tests/unit/lib/test_opensearch_lock.py

```diff
@@ -0,0 +1,200 @@
+        # At this point the leader can take the lock
+        os.environ["JUJU_CONTEXT_ID"] = "new-context-id"
+        assert self.harness.charm.node_lock.acquired
+
+    @responses.activate
+    @patch("socket.socket.connect")
+    def test_node_lock_has_online_nodes_leader_acquire_lock_via_document(self, _):
+        # Initializing mocks showing the unit online in the cluster
+        mock_response_root(self.charm.unit.name, self.charm.opensearch.host)
+        mock_response_nodes(self.charm.unit.name, self.charm.opensearch.host)
+
+        # No unit has the lock at this moment
+        expected_response = {"unit-name": ""}
+
+        responses.add(
+            method="GET",
+            url=f"https://{self.charm.opensearch.host}:9200/.charm_node_lock/0",
+            json=expected_response,
+            status=200,
+        )
+
+        # The leader does not have the lock at this moment, but starts "queuing up"
+        # If the charm code raises an uncaught exception later in the Juju event
+        self.harness.set_leader(is_leader=True)
+        self.harness.update_relation_data(
+            self.lock_rel_id, f"{self.charm.app.name}", {"unit-with-lock": ""}
+        )
+        assert not self.harness.charm.node_lock.acquired
+
+        assert (
+            self.harness.get_relation_data(self.lock_rel_id, f"{self.charm.unit.name}")[
+                "lock-requested"
+            ]
+            == "true"
+        )
+
+        # A new event introduces a new Juju context ID
+        # At this point the leader can take the lock
+        os.environ["JUJU_CONTEXT_ID"] = "new-context-id"
+        assert self.harness.charm.node_lock.acquired
+
+    def test_node_lock_no_online_nodes_depating_node_doesnt_break(self):
+        """Departing unit may be part of the relation while having no entry in the databag.
+
+        # https://github.com/canonical/opensearch-operator/issues/323
+        """
+        # The leader does not have the lock at this moment, but starts "queuing up"
+        # If the charm code raises an uncaught exception later in the Juju event
+        self.harness.set_leader(is_leader=True)
+        self.harness.update_relation_data(
+            self.lock_rel_id, f"{self.charm.app.name}", {"unit-with-lock": ""}
+        )
+
+        # We simulate that a departing node is still part of the relation
+        # however has no more data in the databag
+        unit1 = get_relation_unit(
+            self.harness.model, NodeLockRelationName, f"{self.harness.charm.app.name}/1"
+        )
+        databag = self.harness.charm.model.get_relation(NodeLockRelationName).data
+        databag._data.pop(unit1)
+
+        # The property function executes healthy
+        # instead of breaking over the unit missing from the databag
+        assert not self.harness.charm.node_lock.acquired
+
+    @responses.activate
+    @patch("socket.socket.connect")
+    def test_node_lock_has_online_nodes_depating_node_doesnt_break(self, _):
```

Comment:
"departing"

Reply:
Let me fix the typo in the "neighbor" PR (similar bugfix) (avoiding to loose 2 precious approvals ;-) )